app/src/main/java/org/mozilla/fenix/settings/sitepermissions/SitePermissionsExceptionsFragment.kt

```diff
@@ -58,27 +60,33 @@ class SitePermissionsExceptionsFragment :
         recyclerView = rootView.findViewById(R.id.exceptions)
         recyclerView.layoutManager = LinearLayoutManager(requireContext())
 
+        val adapter = ExceptionsAdapter(this).apply {
+            addLoadStateListener { loadState ->
+                if (loadState.source.refresh is LoadState.NotLoading &&
+                    loadState.append.endOfPaginationReached &&
+                    itemCount < 1
+                ) {
+                    showEmptyListMessage()
+                } else {
+                    hideEmptyListMessage()
+                }
+            }
+        }
+
         viewLifecycleOwner.lifecycleScope.launch(Dispatchers.IO) {
-            val sitePermissionsPaged = requireContext().components.core.permissionStorage.getSitePermissionsPaged()
-
-            withContext(Main) {
-                val adapter = ExceptionsAdapter(this@SitePermissionsExceptionsFragment)
-                val liveData = LivePagedListBuilder(sitePermissionsPaged, MAX_ITEMS_PER_PAGE).build()
-
-                liveData.observe(
-                    viewLifecycleOwner,
-                    {
-                        if (it.isEmpty()) {
-                            showEmptyListMessage()
-                        } else {
-                            hideEmptyListMessage()
-                            adapter.submitList(it)
-                            recyclerView.adapter = adapter
-                        }
-                    }
-                )
+            val dataSourceFactory =
+                requireContext().components.core.permissionStorage.getSitePermissionsPaged()
+
+            val permissions = Pager(PagingConfig(MAX_ITEMS_PER_PAGE), null) {
+                dataSourceFactory.asPagingSourceFactory().invoke()
+            }.flow
+
+            permissions.collect {
+                adapter.submitData(it)
             }
         }
+
+        recyclerView.adapter = adapter
```

Comment:
Can we move this just after the adapter was initialized?

Reply:
Yep, put that into  the`apply` block